XMR protocol optimizations, hardening

ph4r05 opened the pull request:
Protobufs:
- prepare for CLSAG support (required for client side change)
- add subaddress indices for key image refresh

Optimization 1: prepare for permutation sending removal, specify sort index

- related to the #786
- specify source entry ordering index prior sorting by key images as original HMAC keys are generated based on these.
- permutation checked just by valid HMACs, size of the set, key image sort order
- sending permutation is now deprecated, will be removed in the following protocol versions
- more strict state transition checks, guard strict check with respect to steps ordering

Optimization 2: send only required data

- related to the #786
- real_out_additional_tx_keys contains only one element as nothing more is needed during signature
- only `src_entr.outputs[index]` is HMACed and always present. Other outputs are present only if needed which reduces comm and CPU overhead.
- getting rid of subaddresses dictionary (memory requirements), now subaddr indices are present per source entry so keys are computed when needed

New protocol features on the client side: https://github.com/monero-project/monero/pull/6446, GUI: https://github.com/monero-project/monero-gui/pull/2829

I've tested this with https://github.com/ph4r05/monero/releases/tag/v0.15.0.0-tests-u18.04-03 

## Testing
With https://github.com/ph4r05/monero/releases/tag/v0.15.0.0-tests-u18.04-03 

```
# test older client version
TREZOR_CLIENT_VERSION=1 ./trezor_tests --chain_path chain.bin

# test the newest client version
TREZOR_CLIENT_VERSION=3 ./trezor_tests --chain_path chain.bin
```

prusnak commented on common/protob/messages-monero.proto:
```diff
@@ -142,6 +142,7 @@ message MoneroTransactionInitRequest {
         optional uint32 client_version = 13;  // connected client version
         optional uint32 hard_fork = 14;       // transaction hard fork number
         optional bytes monero_version = 15;   // monero software version
+        optional uint32 sig_version = 16;     // MLSAG / CLSAG
```
Is it really necessary to add a new field? Will the new fork support both MLSAG and CLSAG? If not, I think the `sig_version` value could be inferred unambiguously from the `hard_fork` field.

ph4r05 replied:
it is probably not required. Depends on how the hard fork rules will be set when it comes to it. If certain hard fork only permits CLSAG but does not enforces them, then we want to leave the decision on the signing algorithm to the wallets. (Similar situation was with bulletproofs if I remember correctly).

If CLSAG usage will be determined purely by hard fork, the field is not needed.

ph4r05 commented:
@prusnak I could also remove support for old Monero clients so we have only the current and next hard fork support to clean up the codebase. Would you prefer me to do it all in one PR (this) or later after these 2 are merged?

prusnak commented:
I'll let @onvej-sl decide what's better for him.

prusnak commented on core/src/apps/monero/xmr/crypto/__init__.py:
```diff
@@ -304,3 +304,21 @@ def xor8(buff, key):
     return buff
+
+
+def cmp_arr(a, b):
```
Do we really need this function? (Micro)Python can compare arrays and tuples just fine:

```
>>> [1,2,3] <= [1,2,3]
True
>>> b"123" <= b"123"
True
>>> [1,2,3] < [1,2,3]
False
>>> b"123" < b"123"
False
>>> [1,2,3] < [1,2,4]
True
>>> b"123" < b"124"
True
>>> None < [1,2,3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'NoneType' and 'list'
>>> None < b"123"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'NoneType' and 'bytes'
```

The only problem might be if you really require an exception in case the input lengths are different, because that just works fine with comparison operators:

```
>>> [1,2,3] < [1,2,3,4]
True
>>> [1,2,3,4] < [1,2,3]
False
>>> b"123" < b"1234"
True
>>> b"1234" < b"123"
False
```

In this case we might want to add an extra check just for the length.

ph4r05 replied:
there is a problem with byte array

```python
>>> a=bytearray([1,2])
>>> b=bytearray([1,3])
>>> a<b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported types for __lt__: 'bytearray', 'bytearray'
```

prusnak replied:
Ah, I was looking at `get_ki_from_vini` at thought it returns bytes, which is not the case. There is something off with the typing. Maybe we can rewrite the function like this and the comparison will start working?

```
def get_ki_from_vini(vini_bin: bytes) -> bytes:
    return bytes(vini_bin[-32:])
```

ph4r05 replied:
I wanted to avoid memory fragmentation by redundant allocations. However, this is the only usage of the `cmp_arr` function, so implications by reallocating 32bytes to `bytes` should be OK I guess.

ph4r05 replied:
But bytearray was maybe reallocating as well. OK I will do the change.

prusnak replied:
Thanks!

ph4r05 commented:
I will do few more testing with the new protocol features. Then I'll convert PR back from draft.

ph4r05 commented:
I am done with the PR. It is ready for review. 
I've tested this with https://github.com/ph4r05/monero/releases/tag/v0.15.0.0-tests-u18.04-03 

How to test:
```
# test older client version
TREZOR_CLIENT_VERSION=1 ./trezor_tests --chain_path chain.bin

# test the newest client version
TREZOR_CLIENT_VERSION=3 ./trezor_tests --chain_path chain.bin
```

ph4r05 commented on core/src/apps/monero/signing/step_09_sign_input.py:
```diff
@@ -171,13 +187,88 @@ async def sign_input(
+    state.mem_trace(8, True)
+    state.last_step = state.STEP_SIGN
     return MoneroTransactionSignInputAck(
         signature=mg_buffer, pseudo_out=crypto.encodepoint(pseudo_out_c)
```
We changed protobuf so it accepts list of bytes as bytes as well. We need to do it this way to avoid memory fragmentation problems and to have random access to MLSAG signature as it is being generated.

ph4r05 commented:
Thanks for comments. I will go through it over the weekend and adjust the PR. As you mentioned more refactoring I will get rid of the old code which is no longer needed for current hard fork versions.

ph4r05 commented on core/src/apps/monero/signing/offloading_keys.py:
```diff
@@ -114,12 +114,28 @@ def det_comm_masks(key_enc, idx: int) -> Sc25519:
 async def gen_hmac_vini(key, src_entr, vini_bin, idx: int) -> bytes:
```
If I remember correctly, type annotations are dropped in runtime so we don't pay import costs on device, right? If I import it in the `if __debug__:`

ph4r05 replied:
Hm maybe `if False` is better?

onvej-sl replied:
> If I remember correctly, type annotations are dropped in runtime so we don't pay import costs on device, right? If I import it in the if False:

Exactly.

ph4r05 commented:
OK, so I worked on the comments and added type annotations all over the place :D 
It should be quite covered for now. All tests with Monero C++ binary are passing in both client versions variants (older and newer Monero wallet binaries). `make style` passes as well.

Unfortunately, I had to merge also #943 to this commit as such large refactoring made quite hard to keep those changes separated. I hope it does not cause inconvenience as you already managed to review the CLSAG code. Thanks for that!

Ad local state and initializers - I liked this idea so I checked the state variables to figure out which can be separated into local states. I found out that only 2-3 variables like `last_ki` that are not used across different steps. To this extent, it seemed like a bit overhead to separate it to local states (however design-wise it is a nice idea). I don't have much spare time for open source projects so I rather prioritized type hinting as it seemed like technical debt.

ph4r05 commented:
I've also cleaned up a code base a bit, removed old unused features related to old hard forks.

- `state.is_bulletproof_v2()` is removed as it is always true
- `state.is_det_mask()` is removed as it is always true
- `state.bp_version` is removed, we assume version 2, may be changed later if this scheme changes
- `client_version == 0` is not supported anymore (very old binaries, hard forks)
- removed support for old range proofs schemes, serializations, ecdh info serializations
- klrki is removed as multisig support is not yet supported and it is not planned for a long time

ph4r05 commented:
It's ready for review now :)

ph4r05 commented:
@onvej-sl I am thinking about closing CLSAG PR as this has all in it, what do you think?

onvej-sl commented:
> @onvej-sl I am thinking about closing CLSAG PR as this has all in it, what do you think?
Agree.

onvej-sl commented:
Anyway, I spend a couple of hours by adding type annotations. I also decided to refactor a few methods and classes, most importantly:
  * I split `MemoryReaderWriter` into `ByteReader` and `ByteWriter'
  * I changed the interface of the methods in `xmr/crypto/__init__.py`.

It's [here](https://github.com/trezor/trezor-firmware/commits/onvej-sl/xmr-protocol), yet unfinished.

ph4r05 commented:
> Anyway, I spend a couple of hours by adding type annotations. I also decided to refactor a few methods and classes, most importantly:
> 
> * I split `MemoryReaderWriter` into `ByteReader` and `ByteWriter'
> * I changed the interface of the methods in `xmr/crypto/__init__.py`.
> 
> It's [here](https://github.com/trezor/trezor-firmware/commits/onvej-sl/xmr-protocol), yet unfinished.

wow nice! So I will cherry-pick your commit and run all the tests against it, OK? Is it production ready?

onvej-sl commented:
> > Anyway, I spend a couple of hours by adding type annotations. I also decided to refactor a few methods and classes, most importantly:
> > 
> > * I split `MemoryReaderWriter` into `ByteReader` and `ByteWriter'
> > * I changed the interface of the methods in `xmr/crypto/__init__.py`.
> > 
> > It's [here](https://github.com/trezor/trezor-firmware/commits/onvej-sl/xmr-protocol), yet unfinished.
> 
> wow nice! So I will cherry-pick your commit and run all the tests against it, OK? Is it production ready?

Unfortunately, it is not ready, yet. It won't pass the tests.

ph4r05 commented:
@onvej-sl I've added some comments to the code.

I like the typing, but I am a bit worried about changes in the crypto code. `decodepoint_into` should always return the value of destination buffer so its result can be immediately used as an argument. It leads to nicer code.

```python
             C_P = crypto.point_sub(
                 crypto.decodepoint_into(tmp_pt, pubs[i].commitment), C_offset
             )
             crypto.add_keys2_into(
                 L, ss[i], c_p, crypto.decodepoint_into(tmp_pt, pubs[i].dest)
             )
``` 

```python
             C_P = crypto.point_sub(tmp_pt, C_offset)
             crypto.decodepoint_into(tmp_pt, pubs[i].dest)
             crypto.add_keys2_into(L, ss[i], c_p, tmp_pt)
             crypto.scalarmult_into(tmp_pt, C_P, c_c)
             crypto.point_add_into(L, L, tmp_pt)
```

ph4r05 commented:
I would maybe prefer to merge this PR first and then deal with the typing from your commit as the changes are quite huge now. What do you think?

ph4r05 commented:
Another thing - we should investigate the memory impact of the newly added code. The crypto wrappers could increase the footprint as it is imported basically everywhere. However, we should be good after the last BulletProof optimizations. Still, it is better to measure it during tests with a real hardware.

onvej-sl commented:
> I would maybe prefer to merge this PR first and then deal with the typing from your commit as the changes are quite huge now. What do you think?

I agree.

> Another thing - we should investigate the memory impact of the newly added code. The crypto wrappers could increase the footprint as it is imported basically everywhere. However, we should be good after the last BulletProof optimizations. Still, it is better to measure it during tests with a real hardware.

Absolutely.

onvej-sl commented:
> decodepoint_into should always return the value of the destination buffer so its result can be immediately used as an argument. It leads to a nicer code.

I don't quite agree with that. In my opinion the function `def point_add_into(r: Ge25519, a: Ge25519, b: Ge25519) -> Ge25519` is a mixed c-style and python-style function. I don't expect a function that returns the sum of two points to have any side effects. 

> Not returning the values from the new crypto wrappers is probably the reason for tests failing. And it can lead to numerous other bugs.

I went through all occurrences of the functions and fixed that. I don't deny that I may have missed something. I agree that it can lead to numerous bugs. Nevertheless, I hope we will make mypy work on monero soon, which should solve it.

ph4r05 commented:
> I don't quite agree with that. In my opinion the function `def point_add_into(r: Ge25519, a: Ge25519, b: Ge25519) -> Ge25519` is a mixed c-style and python-style function. I don't expect a function that returns the sum of two points to have any side effects.

I am not sure I follow which side effects are you referring to. 
`*_into` functions are essential for complex crypto code as it avoids new memory allocations, otherwise Bulletproof would not be possible. Thus it has to modify the `r`. Moreover, returning the `r` simplifies the usage of the function as it can be easily composed without a need to flatten the code. 

In my opinion, `*_into` functions have to return the buffers because IMO composing the crypto code is clearer and less error-prone. 

I spent tons of hours on the Bulletproof code so it can run on the real hardware and tried numerous options until I got to the current state. I just personally like the current state of the crypto with `*_into` and Bulletproof code much more and it is easier to maintain.

ph4r05 commented:
One way to overcome the "side effect expectation" is to add this principle to the comment. 
I come to the understanding that in the embedded world we have to find tradeoffs due to RAM limits and `_into` functions is one of them. 

I really don't like the flattened code version (hard to maintain, effect of not returning the into buffer) and I don't have time to refactor the major part of the crypto code and check the memory effectiveness again. I would rather avoid changing these things a lot otherwise I would have to try harder to find a motivation to continue supporting this project.

ph4r05 commented:
(I hope I managed to explain my point after refining it in couple of edits iterations :D)

onvej-sl commented:
Big picture: My intention was to add type annotation in order to run mypy on the code. Motivation behind adding mypy is both to avoid some bugs and to write prettier code.

First, I'm not against `*_into` functions. I'm against `*_into` functions that returns a value (via `return` keyword). So this is not a question of speed or memory consumption. I don't consider these functions to have good coding style. For example, I don't like the idea of mixing `*_into` and `*` functions, it can lead to awful statements such as
```
return point_add(a, point_double_into(a, a)) # return 4 * a
```

Just for illustration, here is the list of changes I made. It's not complete, I omitted similar entries. The last one is by far the most common.
```diff
-return (
-    _sc_mul(self.cur, self.cur, self.x)
-    if not self.raw
-    else crypto.sc_mul_into(self.cur, self.cur, self.x)
-)
+if not self.raw:
+    _sc_mul(self.cur, self.cur, self.x)
+else:
+    crypto.sc_mul_into(self.cur, self.cur, self.x)
+return self.cur

-return crypto.encodeint_into(_tmp_bf_0, crypto.decodeint_into(_tmp_sc_1, sc)) == sc
+crypto.decodeint_into(_tmp_sc_1, sc)
+crypto.encodeint_into(_tmp_bf_0, _tmp_sc_1)
+return _tmp_bf_0 == sc

-return crypto.encodepoint_into(dst, self.acc)
+crypto.encodepoint_into(dst, self.acc)
+return dst

-crypto.add_keys2_into(
-    L, ss[j], c_old, crypto.decodepoint_into(Hi, pk[i][j])
-)
+crypto.decodepoint_into(Hi, pk[i][j])
+crypto.add_keys2_into(L, ss[j], c_old, Hi)

-hsh_PC(crypto.encodepoint_into(tmp_bf, sI))
+crypto.encodepoint_into(tmp_bf, sI)
+hsh_PC(tmp_bf)

```

I don't consider the flatten version less legible and I like it more than the previous version.

My only concern is that I missed a place in the code that uses the return value. This will be solved when the type annotation are good enough to use mypy.

Maybe we should discuss it somewhere else not to spam this PR. I also have to postpone the work on the branch because I have something with higher priority now.

onvej-sl commented:
Just a technical note, our team agreed on rules that make reviewing easier:
  * When you want to update a branch that is being reviewed please use fixup commits instead of force pushing the branch.
  * Please don't resolve review comments yourself, let the reviewer to do it.

We will appreciate it.

ph4r05 commented:
> Just a technical note, our team agreed on rules that make reviewing easier:
> 
> * When you want to update a branch that is being reviewed please use fixup commits instead of force pushing the branch.
> * Please don't resolve review comments yourself, let the reviewer to do it.
> 
> We will appreciate it.

OK cool this makes sense. The squash will be performed once the PR is approved?

ph4r05 commented:
ad `*_into` functions: OK so you just object returning the object. I don't see any objective argument about why this should be a bad thing. Ugly code can be written with and without `into` functions. It is maybe highly subjective. 

So we probably agree to disagree. In my opinion, the flattened code is a terrible solution as you lose track of the data flow and you simply cannot see the equation in your head just by looking at the code. There is nontrivial work needed to grasp what is going on in the code. 

I illustrate it on the example I've gave recently:

```
             C_P = crypto.point_sub(
                 crypto.decodepoint_into(tmp_pt, pubs[i].commitment), C_offset
             )
             crypto.add_keys2_into(
                 L, ss[i], c_p, crypto.decodepoint_into(tmp_pt, pubs[i].dest)
             )
```

vs. all smashed line per line. It is obviously harder to read and very easy to miss the bug. 

```
             C_P = crypto.point_sub(tmp_pt, C_offset)
             crypto.decodepoint_into(tmp_pt, pubs[i].dest)
             crypto.add_keys2_into(L, ss[i], c_p, tmp_pt)
             crypto.scalarmult_into(tmp_pt, C_P, c_c)
             crypto.point_add_into(L, L, tmp_pt)
```


If you don't like mixing `into` and non-into functions, we can change all to `into` functions with the first parameter being None. That solves the problem even though I don't see the issue with the mixing. If you see code that is hard to read, refactor it to a simpler version.

Further discussions on this topic can be then moved to a new PR once you resume work on this.

ph4r05 commented on core/src/apps/monero/signing/step_06_set_output.py:
```diff
@@ -410,36 +425,27 @@ def _get_ecdh_info_and_out_pk(state: State, tx_out_key, amount, mask, amount_key
     crypto.sc_add_into(state.sumout, state.sumout, mask)
 
     # masking of mask and amount
```
@onvej-sl do you want me to unresolve these?

ph4r05 commented:
https://github.com/monero-project/monero/pull/6446 has been merged to Monero, https://github.com/monero-project/monero-gui/pull/2829 waits for merge to monero-gui. 

Once this PR is merged and we know the firmware version that contains it I have to issue a nwe PR to the monero so it recognizes the new client version.

ph4r05 commented:
OK so now both monero-project/monero#6446 and monero-project/monero-gui#2829 are merged. 

Is there pls something waiting for my response in this PR?

ph4r05 commented:
I've just checked with the Monero team, there is 1 week before tagging a new major version release 0.16.0.0, so if we want the newest changes to support in that release, we have to know the Trezor firmware version that supports the newest changes, like CLSAG and protocol version 3. 

Do you think we can make it in time? I would need to know the Trezor firmware version containing this PR (without changes on the public API part, cliet protocol) by Wed (13. May 2020). If we make it, new major release will support the newest protocol 3. If not, it will take another 6 months for a new monero client release.

ph4r05 commented:
ping @onvej-sl @prusnak, do you think we can have the firmware version containing this PR or should I leave it and we pass this 0.16.0.0 major release?

prusnak commented:
@ph4r05 The firmware version containing this change will be 2.3.1, please add this to 0.16.0.0 release

onvej-sl commented:
> OK cool, this makes sense. Is this a new thing?

I guess a few month old.

> The squash will be performed once the PR is approved?

Exactly.

> I thought GitHub shows you a difference since the last review even though the code was squashed and force-pushed. Isn't it like this? (this might help with other projects as well)

I wish it would. Maybe it does and I can't see it. :/ 

> Is there pls something waiting for my response in this PR?

I'm sorry, I didn't replied, I was deeply immersed in my work.

prusnak commented:
Thanks! Merging to internal branch to see whether this build in our internal CI.

ph4r05 commented:
> Thanks! Merging to internal branch to see whether this build in our internal CI.

If there goes something wrong pls let me know. 

I will update Monero testing binary (compiled from the monero c++ repo), now it supports specifying which client protocol version should be tested.

rating89us commented:
> I thought GitHub shows you a difference since the last review even though the code was squashed and force-pushed. Isn't it like this? (this might help with other projects as well)

It does show, but you have to click on the first link (red link below) to see the changes of a code that was force-pushed:
![image](https://user-images.githubusercontent.com/45968869/82392466-3839ff00-9a44-11ea-8197-0a6e6260c153.png)

ph4r05 commented:
Btw I've updated the testing release

https://github.com/ph4r05/monero/releases/tag/v0.15.0.0-tests-u18.04-03
